### ISSUE TITLE
ModuleNotFoundError (tests) for coherent.build

### ISSUE THREAD

**Issue by jaraco** [coherent-oss/coherent.test#7]
Running with `coherent.cli 0.2` and `coherent.test 0.5`, the tests fail collection:

```
 @ coh test
============================================================== test session starts ===============================================================
platform darwin -- Python 3.13.0b2, pytest-8.2.2, pluggy-1.5.0
rootdir: /Users/jaraco/code/coherent-oss/coherent.build
configfile: pyproject.toml
plugins: typeguard-4.3.0, coherent.test-0.5.0
collected 12 items / 2 errors                                                                                                                    

===================================================================== ERRORS =====================================================================
______________________________________________________ ERROR collecting tests/test_hooks.py ______________________________________________________
ImportError while importing test module '/Users/jaraco/code/coherent-oss/coherent.build/tests/test_hooks.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../Library/Application Support/pipx/venvs/coherent-cli/lib/python3.13/site-packages/coherent/test/monkey.py:22: in import_path
    return importlib.import_module(
/opt/python/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'coherent.build.tests'
```

Even using the workaround for #3, the output is the same.

The issue, I think, is that `coherent.build` is found in site-packages, but since `tests` are excluded from the install, they're never found.

If, however, I use a modified workaround, the tests run:

```
 @ pip-run -e . coherent.test -- -m coherent.test
```

Because the local project is editable-installed, `tests` _does_ appear in `coherent.build`.

**Comment by johnslavik** [comment]
Indeed the reason for the collection to fail is that test modules aren't installed:

```
tempora on  experiment/coherent 
❯ pipx run coherent.test==0.6.1
===================================================== test session starts ======================================================
platform linux -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0
rootdir: /home/astaroth/Projects/tempora
configfile: pyproject.toml
plugins: pytest_freezer-0.4.8, coherent.test-0.6.1, typeguard-4.3.0
collecting ... 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/coherent/test/monkey.py(26)import_path()
-> return importlib.import_module(
(Pdb) locals()
{'path': PosixPath('/home/astaroth/Projects/tempora/__init__.py'), 'root': PosixPath('/home/astaroth/Projects/tempora'), 'unused_kwargs': {'mode': 'prepend', 'consider_namespace_packages': False}, 'rel_path': PosixPath('__init__'), 'rel_name': '__init__'}
(Pdb) best_name()
'tempora'
(Pdb) import sys
(Pdb) sys.modules['tempora']
<module 'tempora' from '/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/__init__.py'>
(Pdb) sys.modules['tempora'].__file__
'/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/__init__.py'
(Pdb) path = pathlib.Path(sys.modules['tempora'].__file__)
(Pdb) tempora_sitedir = path.parent
(Pdb) [*tempora_sitedir.iterdir()]
[PosixPath('/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/__init__.py'), PosixPath('/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/utc.py'), PosixPath('/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/timing.py'), PosixPath('/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/__pycache__'), PosixPath('/home/astaroth/.cache/pipx/4828c6f12bdd90a/lib/python3.12/site-packages/tempora/schedule.py')]
(Pdb) 
```

**Comment by johnslavik** [comment]
What concerns me is that even when `project_on_path()` installs the project as editable and inserts it into the path, the necessary modules aren't present.

**Comment by jaraco** [comment]
I ran `coh test` on `tempora@experiment/coherent` and it passed:

```
 tempora experiment/coherent 🐚 coh test
=============================================================== test session starts ===============================================================
platform darwin -- Python 3.13.0b2, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/jaraco/code/jaraco/tempora
configfile: pyproject.toml
plugins: pytest_freezer-0.4.8, typeguard-4.3.0, coherent.test-0.5.0
collected 42 items                                                                                                                                

__init__.py ............                                                                                                                    [ 28%]
schedule.py ...                                                                                                                             [ 35%]
tests/test_schedule.py ...................                                                                                                  [ 80%]
tests/test_timing.py ..                                                                                                                     [ 85%]
timing.py .....                                                                                                                             [ 97%]
utc.py .                                                                                                                                    [100%]

=============================================================== 42 passed in 1.41s ================================================================
```

I then reinstalled `coherent.cli` and now I get the same ModuleNotFound errors... so something changed between coherehent.cli 0.3.0 and 0.3.1 or maybe coherent.build (coherent.test was 0.6.1 for both).

**Comment by jaraco** [comment]
I downgraded to coherent.cli 0.3.0 and the problem persists.

**Comment by jaraco** [comment]
This issue may be a duplicate of #3. Here's what's happening with tempora:

In https://github.com/jaraco/jaraco.vcs/commit/4ccb6b0ccdc972ca2da1893ecd58c4ad84f09bb6, jaraco.vcs added a dependency on tempora. Because of the dependency chain: `coherent.test` → `coherent.build` → `jaraco.vcs` → `tempora`, installing `coherent.test` causes a natural install of `tempora` and because importing `coherent.test` implies importing `tempora`, the naturally-installed version is imported and gets precedence before the editable-installed version is made available.

I confirmed by `pipx runpip coherent.cli install 'jaraco.vcs<2.3'` and `pipx runpip coherent.cli uninstall -y tempora`, and now `coh test` works.

I think I just need to solve #3, but I filed this issue separately because there's potentially more going on.
